File: leetcodes/641/solution.cpp

#include <memory>

class Node {
 public:
  Node* next;
  Node* prev;
  int val;

  explicit Node(int val) : val(val), next(nullptr), prev(nullptr) {}
};

class MyCircularDeque {
 public:
  int sizes;
  int curr_size;
  Node* head;
  Node* tail;

  explicit MyCircularDeque(int k) : sizes(k), curr_size(0) {
    head = new Node(-1);
    tail = new Node(-1);

    head->next = tail;
    head->prev = tail;
    tail->next = head;
    tail->prev = head;
  }

  bool insertFront(int value) {
    if (isFull())
      return false;
    _insert(head, new Node(value));
    return true;
  }

  bool insertLast(int value) {
    if (isFull())
      return false;
    _insert(tail->prev, new Node(value));
    return true;
  }

  bool deleteFront() {
    if (isEmpty())
      return false;
    _delete(head->next);
    return true;
  }

  bool deleteLast() {
    if (isEmpty())
      return false;
    _delete(tail->prev);
    return true;
  }

  int getFront() { return head->next->val; }

  int getRear() { return tail->prev->val; }

  bool isEmpty() { return head->next == tail; }

  bool isFull() { return curr_size == sizes; }

 private:
  void _insert(Node* node1, Node* node2) {
    // linking new node
    node2->next = node1->next;
    node2->prev = node1;
    // linking next node to new node
    node1->next->prev = node2;
    // linking prev node to new node
    node1->next = node2;

    curr_size += 1;
  }

  void _delete(Node* node) {
    node->prev->next = node->next;
    node->next->prev = node->prev;

    curr_size -= 1;
  }
};

File: leetcodes/641/solution.py
class Node:
    def __init__(self, val, next_node, prev_node):
        self.val = val
        self.next = next_node
        self.prev = prev_node

class MyCircularDeque(object):

    def __init__(self, k):
        """
        :type k: int
        """
        self.sizes = k
        self.curr_sizes = 0
        self.head = Node(None, None, None)
        self.last = Node(None, self.head, self.head)
        self.head.prev = self.last
        self.head.next = self.last

    def _insert(self, node1, node2):
        node1.next, node2.next, node2.prev, node1.next.next.prev = node2, node1.next, node1, node2

    def _delete(self, node):
        node.prev.next, node.next.prev = node.next, node.prev

    def insertFront(self, value):
        """
        :type value: int
        :rtype: bool
        """
        if (self.curr_sizes == self.sizes):
            return False
        self.curr_sizes += 1
        self._insert(self.head, Node(value,None,None))
        return True
        

    def insertLast(self, value):
        """
        :type value: int
        :rtype: bool
        """
        if (self.curr_sizes == self.sizes):
            return False
        self.curr_sizes += 1
        self._insert(self.last.prev, Node(value,None,None))
        return True
       

    def deleteFront(self):
        """
        :rtype: bool
        """
        if (self.head.next == self.last):
            return False
        self.curr_sizes -= 1
        self._delete(self.head.next)
        return True
        

    def deleteLast(self):
        """
        :rtype: bool
        """
        if (self.last.prev == self.head):
            return False
        self.curr_sizes -= 1
        self._delete(self.last.prev)
        return True
        

    def getFront(self):
        """
        :rtype: int
        """
        return self.head.next.val if self.head.next.val != None else -1
        

    def getRear(self):
        """
        :rtype: int
        """
        return self.last.prev.val if self.last.prev.val != None else -1

    def isEmpty(self):
        """
        :rtype: bool
        """
        return self.head.next == self.last
        

    def isFull(self):
        """
        :rtype: bool
        """
        return self.curr_sizes >= self.sizes
